core/evm.go

```diff
@@ -107,6 +108,46 @@ func GetHashFn(ref *block.Header, chain ChainContext) func(n uint64) common.Hash
 	}
 }
 
+// GetVRFFn returns a GetVRFFn which retrieves header vrf by number
+func GetVRFFn(ref *block.Header, chain ChainContext) func(n uint64) common.Hash {
+	var cache map[uint64]common.Hash
+
+	return func(n uint64) common.Hash {
+		// If there's no hash cache yet, make one
+		if cache == nil {
+			curVRF := common.Hash{}
+			if len(ref.Vrf()) >= 32 {
+				vrfAndProof := ref.Vrf()
+				copy(curVRF[:], vrfAndProof[:32])
+			}
+
+			cache = map[uint64]common.Hash{
+				ref.Number().Uint64(): curVRF,
+			}
+		}
+		// Try to fulfill the request from the cache
+		if hash, ok := cache[n]; ok {
+			return hash
+		}
+		// Not cached, iterate the blocks and cache the hashes
+		for header := chain.GetHeader(ref.ParentHash(), ref.Number().Uint64()-1); header != nil; header = chain.GetHeader(header.ParentHash(), header.Number().Uint64()-1) {
```

Comment:
what is the use case for caching vrf's all previous blocks vs just pulling the requested block number vrf and caching?

Reply:
It’s following the same logic as in getHashFn. I guess it’s safer to trace back blocks rather than directly query by block number.